ldpd/lde.c

```diff
@@ -1533,3 +1566,104 @@ lde_address_list_free(struct lde_nbr *ln)
 		free(lde_addr);
 	}
 }
+
+static void
+lde_del_label_chunk(void *val)
+{
+	free(val);
+}
+static int
+lde_get_label_chunk(void)
+{
+	int ret;
+	uint32_t start, end;
+
+	log_debug("Getting label chunk");
+	ret = lm_get_label_chunk(zclient_sync, 0, CHUNK_SIZE, &start, &end);
+	if (ret < 0)
+	{
+		log_warnx("Error getting label chunk!");
+		close(zclient_sync->sock);
+		zclient_sync->sock = -1;
+		return -1;
+	}
+
+	on_get_label_chunk_response(start, end);
+
+	return 0;
+}
+static void
+lde_label_list_init(void)
+{
+	label_chunk_list = list_new();
+	label_chunk_list->del = lde_del_label_chunk;
+
+	/* get first chunk */
+	while (lde_get_label_chunk() != 0) {
+		fprintf(stderr, "Error getting first label chunk!\n");
+		sleep(1);
+	}
+}
+
+static void
+on_get_label_chunk_response(uint32_t start, uint32_t end)
+{
+	struct label_chunk *new_label_chunk;
+
+	log_debug("Label Chunk assign: %u - %u", start, end);
+
+	new_label_chunk = calloc(1, sizeof(struct label_chunk));
+
+	new_label_chunk->start = start;
+	new_label_chunk->end = end;
+	new_label_chunk->used_mask = 0;
+
+	listnode_add(label_chunk_list, (void *)new_label_chunk);
+
+	/* let's update current if needed */
+	if (!current_label_chunk)
+		current_label_chunk = listtail(label_chunk_list);
+}
+
+static uint32_t
+lde_get_next_label(void)
+{
+	struct label_chunk *label_chunk;
+	uint32_t i, pos, size;
+	uint32_t label = NO_LABEL;
+
+	while (current_label_chunk) {
+		label_chunk = listgetdata(current_label_chunk);
+		if (!label_chunk)
+			goto end;
+
+		/* try to get next free label in currently used label chunk */
+		size = label_chunk->end - label_chunk->start + 1;
+		for (i = 0, pos = 1; i < size; i++, pos <<= 1) {
+			if (!(pos & label_chunk->used_mask)) {
+				label_chunk->used_mask |= pos;
+				label = label_chunk->start + i;
+				goto end;
+			}
+		}
+		current_label_chunk = listnextnode(current_label_chunk);
+	}
+
+end:
+	/* we moved till the last chunk, or were not able to find a label,
+	   so let's ask for another one */
+	if (!current_label_chunk || current_label_chunk == listtail(label_chunk_list)
+		|| label == NO_LABEL) {
+		if (lde_get_label_chunk() != 0)
+			log_warn("%s: Error getting label chunk!", __func__);
+
+	}
+
+	return NO_LABEL;
```

Comment:
You really want to return "label" here instead :)

I decided to investigate why the ANVL-LDP-26.8 test is failing and discovered that this is the culprit (without fixing this no labels are allocated at all).